src/client.ts

```diff
@@ -1,145 +1,164 @@
-import http from 'http';
-
+import fetch, { Response } from 'node-fetch';
 import { IntegrationProviderAuthenticationError } from '@jupiterone/integration-sdk-core';
-
-import { IntegrationConfig } from './types';
+import {
+  IntegrationConfig,
+  StatusError,
+  BambooHRUser,
+  BambooHRFilesResponse,
+  BambooHRFile,
+  BambooHREmployeeDetails,
+  BambooHREmployeesMap,
+} from './types';
 
 export type ResourceIteratee<T> = (each: T) => Promise<void> | void;
 
-// Providers often supply types with their API libraries.
-
-type AcmeUser = {
-  id: string;
-  name: string;
-};
-
-type AcmeGroup = {
-  id: string;
-  name: string;
-  users?: Pick<AcmeUser, 'id'>[];
-};
-
-// Those can be useful to a degree, but often they're just full of optional
-// values. Understanding the response data may be more reliably accomplished by
-// reviewing the API response recordings produced by testing the wrapper client
-// (below). However, when there are no types provided, it is necessary to define
-// opaque types for each resource, to communicate the records that are expected
-// to come from an endpoint and are provided to iterating functions.
-
-/*
-import { Opaque } from 'type-fest';
-export type AcmeUser = Opaque<any, 'AcmeUser'>;
-export type AcmeGroup = Opaque<any, 'AcmeGroup'>;
-*/
-
-/**
- * An APIClient maintains authentication state and provides an interface to
- * third party data APIs.
- *
- * It is recommended that integrations wrap provider data APIs to provide a
- * place to handle error responses and implement common patterns for iterating
- * resources.
- */
 export class APIClient {
-  constructor(readonly config: IntegrationConfig) {}
+  private readonly clientNamespace: string;
+  private readonly clientAccessToken: string;
 
-  public async verifyAuthentication(): Promise<void> {
-    // TODO make the most light-weight request possible to validate
-    // authentication works with the provided credentials, throw an err if
-    // authentication fails
-    const request = new Promise((resolve, reject) => {
-      http.get(
-        {
-          hostname: 'localhost',
-          port: 443,
-          path: '/api/v1/some/endpoint?limit=1',
-          agent: false,
-          timeout: 10,
-        },
-        (res) => {
-          if (res.statusCode !== 200) {
-            reject(new Error('Provider authentication failed'));
-          } else {
-            resolve();
-          }
-        },
-      );
+  constructor(readonly config: IntegrationConfig) {
+    this.clientNamespace = config.clientNamespace;
+    this.clientAccessToken = config.clientAccessToken;
+  }
+
+  private withBaseUri(path: string): string {
+    return `https://api.bamboohr.com/api/gateway.php/${this.clientNamespace}/${path}`;
+  }
+
+  private async request(
+    uri: string,
+    method: 'GET' | 'HEAD' = 'GET',
+  ): Promise<Response> {
+    return await fetch(uri, {
+      method,
```

Comment:
I notice that this client doesn't use paging. Looking at the BambooHR API reference, I didn't see any evidence of using `nextPage` or `pageSize` args in their API calls... did you look to see if they support pagination? This API could have a huge response for some calls to `/users` depending on the company size.

Reply:
Yes, our client doesn't use any pagination but it's because we haven't really seen any mention of pagination on their API. I've just checked their docs again and tried to be creative with some queries myself and haven't really found a way to paginate requests. It is a bit late at the moment though, let us re-check this again tomorrow (just in case), but as it currently stands it looks like their API doesn't support pagination :man_shrugging:

Reply:
Thanks for double-checking, that's good enough for me.